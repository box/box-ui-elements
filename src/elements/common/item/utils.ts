import { FILE_EXTENSIONS } from './constants';

const getFileIconType = (extension = '') => {
    const iconTypes = {};

    Object.entries(FILE_EXTENSIONS).forEach(([iconType, typeExtensions]) => {
        typeExtensions.forEach(typeExtension => {
            iconTypes[typeExtension] = iconType;
        });
    });

    return iconTypes[extension] || 'default';
};

export { getFileIconType };
