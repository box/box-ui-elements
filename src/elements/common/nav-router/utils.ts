import { History } from 'history';
import { WithNavRouterProps } from './types';

export interface MatchResult {
    isExact: boolean;
    params: Record<string, string>;
}

export function matchPath(pathname: string, path: string | string[]): MatchResult {
    const paths = Array.isArray(path) ? path : [path];
    const params: Record<string, string> = {};

    for (const currentPath of paths) {
        const pathSegments = currentPath.split('/').filter(Boolean);
        const pathnameSegments = pathname.split('/').filter(Boolean);

        if (pathSegments.length <= pathnameSegments.length) {
            let isMatch = true;
            const segmentLength = pathSegments.length;

            for (let i = 0; i < segmentLength; i += 1) {
                const pathSegment = pathSegments[i];
                const pathnameSegment = pathnameSegments[i];

                if (pathSegment.startsWith(':')) {
                    // This is a parameter
                    const paramName = pathSegment.slice(1);
                    params[paramName] = pathnameSegment;
                } else if (pathSegment === '*') {
                    // Wildcard matches anything
                    params[i.toString()] = pathnameSegments.slice(i).join('/');
                    break;
                } else if (pathSegment !== pathnameSegment) {
                    isMatch = false;
                    break;
                }
            }

            if (isMatch) {
                return {
                    isExact: pathSegments.length === pathnameSegments.length,
                    params,
                };
            }
        }
    }

    return { isExact: false, params: {} };
}

export function useMatch(
    props: Partial<WithNavRouterProps> & { history: History },
    path: string | string[],
): MatchResult {
    const location = props.location || props.history.location;
    return matchPath(location.pathname, path);
}
