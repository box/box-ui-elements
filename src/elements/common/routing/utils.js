/**
 * @file Utility functions for routing
 */

/**
 * @typedef {Object} MatchOptions
 * @property {string} path
 * @property {boolean} [exact]
 * @property {boolean} [strict]
 */

/**
 * @typedef {Object} Match
 * @property {string} path
 * @property {string} url
 * @property {boolean} isExact
 * @property {Object.<string, string>} params
 */

/**
 * Matches a pathname against a path pattern
 * @param {string} pathname
 * @param {MatchOptions} options
 * @returns {Match|null}
 */
export const matchPath = (pathname, { path, exact = false, strict = false }) => {
    // Handle array of paths
    if (Array.isArray(path)) {
        for (const p of path) {
            const match = matchPath(pathname, { path: p, exact, strict });
            if (match) return match;
        }
        return null;
    }

    const trimmedPath = String(path || '').replace(/\/$/, '');
    const trimmedPathname = String(pathname || '').replace(/\/$/, '');

    // Extract params from path pattern
    const paramNames = [];
    const regexPath = trimmedPath.replace(/:([^/]+)/g, (match, paramName) => {
        paramNames.push(paramName);
        // Allow commas in parameter values for metadata template IDs
        return paramName === 'filteredTemplateIds' ? '([^/]+)' : '([^/,]+)';
    });

    const regex = new RegExp(`^${regexPath}${strict ? '' : '/?'}${exact ? '$' : ''}`);
    const match = trimmedPathname.match(regex);

    if (!match) {
        return null;
    }

    const params = {};
    paramNames.forEach((name, index) => {
        const value = match[index + 1];
        // Handle special case for filteredTemplateIds which can be comma-separated
        params[name] = value;
    });

    return {
        path,
        url: match[0],
        isExact: match[0] === trimmedPathname,
        params,
    };
};

/**
 * Generates a path with parameters replaced by their values
 * @param {string} path - The path pattern with parameter placeholders
 * @param {Object} [params] - The parameter values to substitute
 * @returns {string} The generated path
 */
export const generatePath = (path, params) => {
    if (!params) {
        return path;
    }

    return Object.keys(params).reduce((acc, key) => {
        const value = params[key];
        const pattern = `:${key}`;
        return acc.replace(pattern, value);
    }, path);
};
