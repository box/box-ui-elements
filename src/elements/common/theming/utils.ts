import isObject from 'lodash/isObject';
import kebabCase from 'lodash/kebabCase';

const convertTokensToCustomProperties = (tokens = {}, prefix = '') => {
    const customProperties = {};

    Object.entries(tokens).forEach(([level, value]) => {
        const levelName = `${prefix}${kebabCase(level)}`;

        if (isObject(value)) {
            const properties = convertTokensToCustomProperties(value, `${levelName}-`);
            Object.entries(properties).forEach(([tokenName, tokenValue]) => {
                customProperties[tokenName] = tokenValue;
            });
        } else {
            customProperties[`--${levelName}`] = value;
        }
    });

    return customProperties;
};

export { convertTokensToCustomProperties };
