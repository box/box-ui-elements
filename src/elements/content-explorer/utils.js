// @flow
import getProp from 'lodash/get';
import { REPRESENTATIONS_RESPONSE_SUCCESS, REPRESENTATIONS_RESPONSE_VIEWABLE } from '../../constants';
import type { BoxItem } from '../../common/types/core';

export default (item: BoxItem) => {
    const status = getProp(item, 'representations.entries[0].status.state');
    return status === REPRESENTATIONS_RESPONSE_SUCCESS || status === REPRESENTATIONS_RESPONSE_VIEWABLE;
};
