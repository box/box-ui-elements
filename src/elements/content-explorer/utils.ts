import React, { useMemo } from 'react';
import { useIntl } from 'react-intl';
import isNil from 'lodash/isNil';
import xor from 'lodash/xor';

import {
    MULTI_VALUE_DEFAULT_OPTION,
    MULTI_VALUE_DEFAULT_VALUE,
    type MetadataTemplate,
    type MetadataFormFieldValue,
} from '@box/metadata-editor';
import type { MetadataFieldType } from '@box/metadata-view';
import type { Selection } from 'react-aria-components';
import type { BoxItem, Collection } from '../../common/types/core';

import messages from '../common/messages';

// Specific type for metadata field value in the item
// Note: Item doesn't have field value in metadata object if that field is not set, so the value will be undefined in this case
type ItemMetadataFieldValue = string | number | Array<string> | null | undefined;

// Get selected item text
export function useSelectedItemText(currentCollection: Collection, selectedItemIds: Selection): string {
    const { formatMessage } = useIntl();

    return useMemo(() => {
        const selectedCount = selectedItemIds === 'all' ? currentCollection.items.length : selectedItemIds.size;
        if (selectedCount === 0) return '';

        // Case 1: Single selected item - show item name
        if (selectedCount === 1) {
            const selectedKey =
                selectedItemIds === 'all' ? currentCollection.items[0].id : selectedItemIds.values().next().value;
            const selectedItem = currentCollection.items.find(item => item.id === selectedKey);
            return selectedItem?.name ?? '';
        }

        // Case 2: Multiple selected items - show count
        return formatMessage(messages.numFilesSelected, { numSelected: selectedCount });
    }, [currentCollection.items, formatMessage, selectedItemIds]);
}

// Check if the field value is empty.
// Note: 0 doesn't represent empty here because of float type field
export function isEmptyValue(value: ItemMetadataFieldValue) {
    if (isNil(value)) {
        return true;
    }

    // date, string, enum
    if (value === '') {
        return true;
    }

    // multiSelect
    if (Array.isArray(value) && value.length === 0) {
        return true;
    }

    // float
    if (Number.isNaN(value)) {
        return true;
    }

    return false;
}

// Check if the field values are equal based on the field types
export function areFieldValuesEqual(value1: ItemMetadataFieldValue, value2: ItemMetadataFieldValue) {
    if (isEmptyValue(value1) && isEmptyValue(value2)) {
        return true;
    }

    // Handle multiSelect arrays comparison
    if (Array.isArray(value1) && Array.isArray(value2)) {
        return xor(value1, value2).length === 0;
    }

    return value1 === value2;
}

// Return default form value by field type
function getDefaultValueByFieldType(fieldType: MetadataFieldType) {
    if (fieldType === 'date' || fieldType === 'enum' || fieldType === 'float' || fieldType === 'string') {
        return '';
    }
    if (fieldType === 'multiSelect') {
        return [];
    }
    return undefined;
}

// Set the field value in Metadata Form based on the field type
function getFieldValue(fieldType: MetadataFieldType, fieldValue: ItemMetadataFieldValue) {
    if (isNil(fieldValue)) {
        return getDefaultValueByFieldType(fieldType);
    }
    return fieldValue;
}

// Check if the field value in Metadata Form is multi-values such as "Multiple values"
export function isMultiValuesField(fieldType: MetadataFieldType, fieldValue: MetadataFormFieldValue) {
    if (fieldType === 'multiSelect') {
        return Array.isArray(fieldValue) && fieldValue.length === 1 && fieldValue[0] === MULTI_VALUE_DEFAULT_VALUE;
    }
    if (fieldType === 'enum') {
        return fieldValue === MULTI_VALUE_DEFAULT_VALUE;
    }
    return false;
}

// Get template instance based on metadata template and selected items
export function useTemplateInstance(metadataTemplate: MetadataTemplate, selectedItems: BoxItem[], isEditing: boolean) {
    const { formatMessage } = useIntl();
    const { displayName, fields, hidden, id, scope, templateKey, type } = metadataTemplate;

    const selectedItemsFields = fields.map(
        ({ displayName: fieldDisplayName, hidden: fieldHidden, id: fieldId, key, options, type: fieldType }) => {
            const defaultItemField = {
                displayName: fieldDisplayName,
                hidden: fieldHidden,
                id: fieldId,
                key,
                options,
                type: fieldType,
                value: getFieldValue(fieldType as MetadataFieldType, undefined),
            };

            const firstSelectedItem = selectedItems[0];
            const firstSelectedItemFieldValue = firstSelectedItem.metadata[scope][templateKey][key];

            // Case 1: Single selected item
            if (selectedItems.length <= 1) {
                return {
                    ...defaultItemField,
                    value: firstSelectedItemFieldValue,
                };
            }

            // Case 2.1: Multiple selected items, but all have the same initial value
            const allItemsHaveSameInitialValue = selectedItems.every(selectedItem =>
                areFieldValuesEqual(selectedItem.metadata[scope][templateKey][key], firstSelectedItemFieldValue),
            );

            if (allItemsHaveSameInitialValue) {
                return {
                    ...defaultItemField,
                    value: getFieldValue(fieldType as MetadataFieldType, firstSelectedItemFieldValue),
                };
            }

            // Case 2.2: Multiple selected items, but some have different initial values
            // Case 2.2.1: Edit Mode
            if (isEditing) {
                let fieldValue = getFieldValue(fieldType as MetadataFieldType, undefined);
                // Add MultiValue Option if the field is multiSelect or enum
                if (fieldType === 'multiSelect' || fieldType === 'enum') {
                    fieldValue = fieldType === 'enum' ? MULTI_VALUE_DEFAULT_VALUE : [MULTI_VALUE_DEFAULT_VALUE];
                    const multiValueOption = options?.find(option => option.key === MULTI_VALUE_DEFAULT_VALUE);
                    if (!multiValueOption) {
                        options?.push(MULTI_VALUE_DEFAULT_OPTION);
                    }
                }
                return {
                    ...defaultItemField,
                    value: fieldValue,
                };
            }

            /**
             * Case: 2.2.2 View Mode
             *
             * We want to show "Multiple values" label for multiple dates across files selection.
             * We use fragment here to bypass check in shared feature.
             * This feature tries to parse string as date if the string is passed as value.
             */
            const multipleValuesText = formatMessage(messages.multipleValues);
            return {
                ...defaultItemField,
                value: React.createElement(React.Fragment, null, multipleValuesText),
            };
        },
    );

    return {
        canEdit: true,
        displayName,
        hidden,
        id,
        fields: selectedItemsFields,
        scope,
        templateKey,
        type,
    };
}
