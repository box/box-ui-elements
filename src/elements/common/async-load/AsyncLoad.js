// @flow
import * as React from 'react';
import AsyncError from './AsyncError';
import { retryNumOfTimes } from '../../../utils/function';

type Props = {
    errorComponent?: React.ComponentType<any>,
    fallback?: React.Element<any>,
    loader: () => Promise<any>,
};

const DEFAULT_NUM_TIMES = 3;
const DEFAULT_INITIAL_DELAY = 500;
const DEFAULT_BACKOFF_FACTOR = 2;

const AsyncLoad = ({ errorComponent, fallback, loader }: Props = {}) => {
    const lazyRetry = () =>
        retryNumOfTimes(
            (successCallback, errorCallback) =>
                loader()
                    .then(successCallback)
                    .catch(errorCallback),
            DEFAULT_NUM_TIMES,
            DEFAULT_INITIAL_DELAY,
            DEFAULT_BACKOFF_FACTOR,
        );
    const LazyComponent = React.lazy(() => loader().catch(lazyRetry));

    return React.forwardRef<Object, React.Ref<any>>((props: Object, ref: React.Ref<any>) => (
        <AsyncError component={errorComponent}>
            <React.Suspense fallback={fallback || null}>
                <LazyComponent ref={ref} {...props} />
            </React.Suspense>
        </AsyncError>
    ));
};

export default AsyncLoad;
