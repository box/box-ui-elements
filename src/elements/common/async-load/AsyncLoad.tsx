import * as React from 'react';
import AsyncError from './AsyncError';
import { retryNumOfTimes } from '../../../utils/function';

interface Props {
    errorComponent?: React.ComponentType<{ error: Error }>;
    fallback?: React.ReactElement;
    loader: () => Promise<{ default: React.ComponentType<unknown> }>;
}

const DEFAULT_NUM_TIMES = 3;
const DEFAULT_INITIAL_DELAY = 500;
const DEFAULT_BACKOFF_FACTOR = 2;

const AsyncLoad = ({ errorComponent, fallback, loader }: Props) => {
    const lazyRetry = () =>
        retryNumOfTimes(
            (successCallback: (value: unknown) => void, errorCallback: (reason: unknown) => void) =>
                loader().then(successCallback).catch(errorCallback),
            DEFAULT_NUM_TIMES,
            DEFAULT_INITIAL_DELAY,
            DEFAULT_BACKOFF_FACTOR,
        );
    const LazyComponent = React.lazy(() => loader().catch(lazyRetry));

    return React.forwardRef((props: Record<string, unknown>, ref: React.Ref<unknown>) => (
        <AsyncError component={errorComponent}>
            <React.Suspense fallback={fallback || null}>
                <LazyComponent ref={ref} {...props} />
            </React.Suspense>
        </AsyncError>
    ));
};

export default AsyncLoad;
