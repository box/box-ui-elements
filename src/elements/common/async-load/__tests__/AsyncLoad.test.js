import * as React from 'react';
import { shallow } from 'enzyme';
import { retryNumOfTimes } from '../../../../utils/function';
import AsyncLoad from '../AsyncLoad';

jest.mock('../../../../utils/function', () => ({
    retryNumOfTimes: jest.fn(),
}));

describe('elements/common/async-load/AsyncLoad', () => {
    const getAsyncComponent = (props = { loader: jest.fn() }) => AsyncLoad(props);
    const getWrapper = (AsyncComponent, props) => shallow(<AsyncComponent {...props} />).dive();

    test('should load the lazy component', () => {
        const AsyncComponent = getAsyncComponent();
        const wrapper = getWrapper(AsyncComponent, { foo: 'bar' });
        expect(wrapper).toMatchSnapshot();
    });

    test('should not invoke the loader until component mounted', () => {
        expect(retryNumOfTimes).not.toHaveBeenCalled();
    });
});
