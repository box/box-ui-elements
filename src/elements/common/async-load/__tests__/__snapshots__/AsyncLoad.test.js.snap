// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/common/async-load/AsyncLoad should load the lazy component 1`] = `
<Suspense
  component={[Function]}
  fallback={null}
>
  <lazy
    foo="bar"
  />
</Suspense>
`;
